cloud/ec2_fleet.go

```diff
@@ -0,0 +1,538 @@
+package cloud
+
+import (
+	"context"
+	"encoding/base64"
+	"fmt"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/credentials"
+	"github.com/aws/aws-sdk-go/service/ec2"
+	"github.com/evergreen-ci/evergreen"
+	"github.com/evergreen-ci/evergreen/model/event"
+	"github.com/evergreen-ci/evergreen/model/host"
+	"github.com/evergreen-ci/evergreen/util"
+	"github.com/mitchellh/mapstructure"
+	"github.com/mongodb/grip"
+	"github.com/mongodb/grip/message"
+	"github.com/pkg/errors"
+)
+
+type ec2FleetManager struct {
+	client      AWSClient
+	credentials *credentials.Credentials
+	settings    *evergreen.Settings
+}
+
+func (m *ec2FleetManager) GetSettings() ProviderSettings {
+	return &EC2ProviderSettings{}
+}
+
+func (m *ec2FleetManager) Configure(ctx context.Context, settings *evergreen.Settings) error {
+	if m.client == nil {
+		m.client = &awsClientImpl{}
+	}
+
+	m.settings = settings
+
+	if settings.Providers.AWS.EC2Key == "" || settings.Providers.AWS.EC2Secret == "" {
+		return errors.New("AWS ID and Secret must not be blank")
+	}
+
+	m.credentials = credentials.NewStaticCredentialsFromCreds(credentials.Value{
+		AccessKeyID:     settings.Providers.AWS.EC2Key,
+		SecretAccessKey: settings.Providers.AWS.EC2Secret,
+	})
+
+	return nil
+}
+
+func (m *ec2FleetManager) SpawnHost(ctx context.Context, h *host.Host) (*host.Host, error) {
+	if h.Distro.Provider != evergreen.ProviderNameEc2Fleet {
+		return nil, errors.Errorf("Can't spawn instance for distro %s: provider is %s",
+			h.Distro.Id, h.Distro.Provider)
+	}
+
+	ec2Settings := &EC2ProviderSettings{}
+	if h.Distro.ProviderSettings != nil {
+		if err := mapstructure.Decode(h.Distro.ProviderSettings, ec2Settings); err != nil {
+			return nil, errors.Wrapf(err, "Error decoding params for distro %s: %+v", h.Distro.Id, ec2Settings)
+		}
+	}
+	grip.Debug(message.Fields{
+		"message":   "mapstructure comparison",
+		"input":     *h.Distro.ProviderSettings,
+		"output":    *ec2Settings,
+		"inputraw":  fmt.Sprintf("%#v", *h.Distro.ProviderSettings),
+		"outputraw": fmt.Sprintf("%#v", *ec2Settings),
+	})
+	if err := ec2Settings.Validate(); err != nil {
+		return nil, errors.Wrapf(err, "Invalid EC2 settings in distro %s: and %+v", h.Distro.Id, ec2Settings)
+	}
+
+	if ec2Settings.AWSKeyID != "" {
```

Comment:
Is any of this duplicated between providers in a way that makes it a candidate for refactoring into a function?

Reply:
refactored some.